Code in order:
1. Filtering methylation  data for CpGs
2. Quality control of methylation data files
3. Wgbstools - genome loading, segmentation, find markers
4. Histograms of segment lengths
5. Clean columns function - removes the “CpG” or “bp” from the numerical value 
6. Oligo markers comparison - Basic characteristics of markers
7. Methylation status and counts of markers (oligo)
8.Mean methylation plots (oligo)
9. Mann-Whitney Tests
10. Adjusting gene association data
11. Input for GREAT database
12. Gene enrichment plots
13. Dendrogram for visualisation of methylation variance between cell-types
14. Marker filtering (500)
15. Markers methylation plots
16. Deriving meth and depth per marker region
17. Building CelFiE input
18. CelFiE output analysis

